### PR TITLE
Add a way to set PlayerControlView fullscreen button state

### LINKED ISSUE #1590: Enable the developer to set the default state of isFullscreen
### Use case description
When multiple video players views are chained, if the user presses the fullscreen button in the ``PlayerControlView`` (the state of the button would go from ``fullScreenButton`` to ``minimalFullScreenButton``) and then navigates to a second ``PlayerControlView`` while still being in fullscreen, the button will be in its default position (``fullScreenButton``) in this new ``PlayerControlView``.

### Proposed solution
Create a public facing method in ``PlayerView`` called ``setIsFullscreen`` which would call ``PlayerControlView.updateFullScreenButtonForState`` with the new state (also add the corresponding ``fullScreenButton`` and would update ``isFullscreen``)

### PR DESCRIPTION
**Pull request by theskyblockman**
This fixes #1590.

I added and documented the new public-facing methods, moved some code to a new private method and used it to not duplicate code.

### CODE REVIEW

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java [hunk 1]
```diff
@@ -1138,6 +1138,16 @@ public void setFullscreenButtonClickListener(@Nullable FullscreenButtonClickList
     controller.setOnFullScreenModeChangedListener(componentListener);
   }
 
+  /**
+   * Sets whether the fullscreen button is fullScreenButton or minimalFullScreenButton.
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java:1142]: Please see the difference between `fullScreenButton` and `minimalFullScreenButton`. They are not the same button. Both buttons act as fullscreen toggles and hence can have exit/enter fullscreen icons

Screenshot 2024-09-04 at 00 01 56

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java [hunk 2]
```diff
@@ -1138,6 +1138,16 @@ public void setFullscreenButtonClickListener(@Nullable FullscreenButtonClickList
...
+   *
+   * @param isFullscreen Whether the player is currently in fullscreen.
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java:1144]: Fullscreen state is semantically inverse to the button display.

* When the player is fullscreen -> button should show "exit fullscreen"
* When the player is minimized -> button should show "enter fullscreen"

It might be best to document this method in terms of the icon drawable you will be setting (since you can't really claim to be imperatively setting the player to a fullscreen activity)

* exo_styled_controls_fullscreen_enter
* exo_styled_controls_fullscreen_exit

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java [hunk 3]
```diff
@@ -1138,6 +1138,16 @@ public void setFullscreenButtonClickListener(@Nullable FullscreenButtonClickList
...
+   */
+  public void setFullscreenButtonState(boolean isFullscreen) {
+    Assertions.checkStateNotNull(controller);
+    controller.updateIsFullScreen(isFullscreen);
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java:1148]: Just a stylistic note - you seem to be mixing `Fullscreen` and `FullScreen`. I understand that a mixture of spellings are currently used throughout the repo, but let's align on not capitalizing "S", since "fullscreen" is one word

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java [hunk 4]
```diff
@@ -1574,14 +1574,25 @@ private void seekToTimeBarPosition(Player player, long positionMs) {
...
+
+  /**
+   * Updates whether the controller is in full screen to its icon
+   * and reports it to to the listener.
+   * @param isFullScreen If the view is in full screen.
+   */
+  public void updateIsFullScreen(boolean isFullScreen) {
+    boolean fullScreenModeChanged = this.isFullScreen != isFullScreen;
+
+    if(fullScreenModeChanged && onFullScreenModeChangedListener == null) {
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java:1588]: I think this line can be simplified to `if(onFullScreenModeChangedListener == null)`.

No matter whether the mode changed or not, if the listener is not there, we don't do anything

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java [hunk 5]
```diff
@@ -1574,14 +1574,25 @@ private void seekToTimeBarPosition(Player player, long positionMs) {
...
 
   private void onFullScreenButtonClicked(View v) {
-    if (onFullScreenModeChangedListener == null) {
+    updateIsFullScreen(!isFullScreen);
+  }
+
+  /**
+   * Updates whether the controller is in full screen to its icon
+   * and reports it to to the listener.
+   * @param isFullScreen If the view is in full screen.
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java:1583]: I would reword this javadoc (see PlayerView comment)

**Review by oceanjules — COMMENTED** [review]
Thanks for a PR!

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java (hunk 4 above)
Comment (icbaker) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java:1588]: Don't we want to update the internal state whether or not we have a listener attached?

Consider:

```java
playerView.setFullscreenButtonState(true);
playerView.setFullscreenButtonClickListener(isFullScreen -> ...);
```

Registering the listener means the button is shown, but the "enter full screen" button will be shown even though we've been told the content is full screen above, and so the "exit full screen" **should** be shown.

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java (hunk 1 above)
Comment (theskyblockman) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java]: I rephrased in 2f1b08fa the javadoc:
Sets whether the player is currently in fullscreen, this will change the displayed icon.

If isFullscreen is true, ``exo_styled_controls_fullscreen_exit`` will be displayed or else ``exo_styled_controls_fullscreen_enter``.

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java (hunk 2 above)
Comment (theskyblockman) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerView.java:1147]: I have rephrased this javadoc too in 2f1b08fa:
Updates whether the controller is in fullscreen, changing its fullscreen icon and reports it to to the listener.

For isFullscreen equals true the icon will be set to ``exo_styled_controls_fullscreen_exit`` or else ``exo_styled_controls_fullscreen_enter``.

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java (hunk 5 above)
Comment (theskyblockman) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java]: I updated this javadoc too:
Updates whether the controller is in fullscreen, changing its fullscreen icon and reports it to to the listener.

For isFullscreen equals true the icon will be set to ``exo_styled_controls_fullscreen_exit`` or else ``exo_styled_controls_fullscreen_enter``.

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java (hunk 4 above)
Comment (oceanjules) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java]: Sorry, yes, you are right - got confused with the previous impl of:

```
  private void onFullScreenButtonClicked(View v) {
    if (onFullScreenModeChangedListener == null) {  <-- can't be true, the button is not visible if the listener==null
      return;
    }
   ...
```

So I guess this line could be changed to `if (!fullScreenModeChanged)` to exit early?

File: libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java [hunk 6]
```diff
@@ -1574,26 +1574,31 @@ private void seekToTimeBarPosition(Player player, long positionMs) {
...
+   *
+   * @param isFullscreen If the view is in full screen.
    */
-  public void updateIsFullScreen(boolean isFullScreen) {
-    boolean fullScreenModeChanged = this.isFullScreen != isFullScreen;
+  public void updateIsFullscreen(boolean isFullscreen) {
+    boolean fullscreenModeChanged = this.isFullScreen != isFullscreen;
 
-    if(fullScreenModeChanged && onFullScreenModeChangedListener == null) {
+    if (fullscreenModeChanged && onFullScreenModeChangedListener == null) {
```
Comment (oceanjules) [inline libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java:1593]: @theskyblockman, please see mine and icbaker's convo above:

tldr; 
* `if (!fullScreenModeChanged)` to exit quickly
* whether or not we have the listener is not important at this point
* If the mode DID change, we still want to update the state and go through the rest of this function

Comment (theskyblockman) [reply libraries/ui/src/main/java/androidx/media3/ui/PlayerControlView.java]: Changed in 616e17af.

**Comment by oceanjules** [comment]
I'm going to send this for internal review now. You may see some more commits being added as I make changes in response to review feedback. Please refrain from pushing any more substantive changes as it will complicate the internal review - thanks!

**Comment by theskyblockman** [comment]
Thank you!
